_posts/2020-10-28-push-it-to-the-limit.md

```diff
@@ -0,0 +1,58 @@
+
+![original moon data format](/assets/moon_data_v1.PNG) ![minified moon data format](/assets/moon_data_v2.PNG)
+
+These changes took our data from 350 bytes down to 200, even if the communication overhead was doubling our data usage we should still see a 10-15% reduction, keeping us under our limit for probably another year of user growth easily. On May 3rd I flipped the switch and patiently waited for the metrics to roll in, and in one week I planned on disabling the old data endpoint entirely.
+
+![bandwidth usage chart over time](/assets/BlogBandwidthNoChange.PNG)
+
+And... no measurable change occurred over the next several days. Perhaps most of our users weren't on the new codebase? On the 9th I killed the old data endpoint, still nothing. While I still didn't know exactly how bad the communication overhead was I'd proven that no change to our data format was going to resolve the issue, we needed to be smarter about requesting data.
+
+## Well That Was Easy
+
+Perhaps not surprisingly, the correct solution really wasn't that difficult to implement. We had decent estimates on when the next event should occur, why don't we trust those estimates a little more? They may be a few minutes off over a 3 hour window, but they're never going to be an hour off. We changed our code to instead of blindly polling once a minute, to poll at the greater of either 80% of the time until the next event, or 1 minute. We'd keep the same accuracy, but reduce the number of requests dramatically. On may 16th I made this change and pushed it out to the users.
+
+![bandwidth usage bar chart over time](/assets/BlogBandwidthChange.PNG)
+
+It was at this time as I was reveling in the huge performance gains that I discovered that we could also view a count of every attempted access to our data via the rule monitoring logic.
+
+![database rules interactions](/assets/BlogRulesChange.PNG)
+
+Using these counts it became easy to calculate that our per communication size averaged almost 4KB, it's not surprising that shaving 150 bytes off this got completely lost in the noise. I also found it interesting to look at the rejection rate, those are the number of users running out of date code still trying to access the old data endpoint after it was disabled. If you want to talk about long tails we were still getting about 6k failed communications daily even into October. October 10th they all finally stopped, leaving me to believe it was one user running ~3 characters 24/7 who hadn't updated in over half a year.
+
+Looking back on things I think we dealt with the problem appropriately when it arose. I can't say the same for when we made the initial design decision. It seems we erred too far towards avoiding premature optimization and didn't fully understand how hard, or easy, it would be to implement the more robust solution instead of going with the easiest approach.
```

Comment:
Hindsight is 20/20. We made the best decision with the information we had at the time, weren't aware that overhead counted (which didn't matter for several _years_ anyway), and had a bajillion things competing for our limited dev time. I consider this one a success.